components/url/lib.rs

```diff
@@ -109,16 +131,32 @@ impl ServoUrl {
         Arc::make_mut(&mut self.0)
     }
 
-    pub fn set_username(&mut self, user: &str) -> Result<(), ()> {
-        self.as_mut_url().set_username(user)
+    pub fn set_username(&mut self, user: &str) -> Result<(), UrlError> {
+        self.as_mut_url()
+            .set_username(user)
+            .map_err(|_| UrlError::SetUsername)
+    }
+
+    pub fn set_ip_host(&mut self, addr: IpAddr) -> Result<(), UrlError> {
+        self.as_mut_url()
+            .set_ip_host(addr)
+            .map_err(|_| UrlError::SetIpHost)
+    }
+
+    pub fn set_password(&mut self, pass: Option<&str>) -> Result<(), UrlError> {
+        self.as_mut_url()
+            .set_password(pass)
+            .map_err(|_| UrlError::SetPassword)
     }
 
-    pub fn set_ip_host(&mut self, addr: IpAddr) -> Result<(), ()> {
-        self.as_mut_url().set_ip_host(addr)
+    pub fn to_file_path(&self) -> Result<::std::path::PathBuf, UrlError> {
+        self.0.to_file_path().map_err(|_| UrlError::ToFilePath)
     }
 
-    pub fn set_password(&mut self, pass: Option<&str>) -> Result<(), ()> {
-        self.as_mut_url().set_password(pass)
+    pub fn from_file_path<P: AsRef<Path>>(path: P) -> Result<Self, UrlError> {
+        Url::from_file_path(path)
+            .map(Self::from_url)
+            .map_err(|_| UrlError::FromFilePath)
```

Comment:
Let's keep the methods in the same order. I'm not sure it makes sense to organize them by their return value.